### DISCUSSION TITLE
Counting different object classes in a separate counter

### DISCUSSION THREAD

**Discussion by Jones034** [ultralytics/ultralytics#8574] [Q&A]
[Counting.zip](https://github.com/ultralytics/ultralytics/files/14458495/Counting.zip)
Hello,
what possibilities do I have to make changes in the attached code so that I do not have one counter but a separate one for each of the two object classes to be counted? One direction is sufficient. How can I export this counted value afterwards?

**Comment by glenn-jocher** [comment]
Hello! 👋

To count different object classes separately, you can use a dictionary where keys are class names and values are counts. After detection, iterate through results and update counts based on detected classes. Here's a simplified example:

```python
from collections import defaultdict

# Assuming 'results' is your detection output
class_counts = defaultdict(int)
for det in results.xyxy[0]:  # detections for first image in batch
    class_id = int(det[-1])  # Get class ID
    class_name = results.names[class_id]  # Convert ID to name
    class_counts[class_name] += 1

print(class_counts)
```

To export these counts, you can write them to a file:

```python
with open('counts.txt', 'w') as f:
    for class_name, count in class_counts.items():
        f.write(f"{class_name}: {count}\n")
```

This will create a `counts.txt` file with each class and its count. Hope this helps!

**Reply by Jones034** [reply]
Thank you very much for your answer. But could not realize what I am looking for

**Reply by glenn-jocher** [reply]
Hi there! 😊

I'm sorry to hear you're still having trouble. Let's try a different approach. If you're looking to count specific object classes detected by YOLOv8, you can do so by iterating through the detection results and keeping track of each class count. Here's a quick example:

```python
from collections import Counter

# Assuming 'results' contains your model's output
class_ids = [int(x[-1]) for x in results.xyxy[0]]
class_counts = Counter(class_ids)

# To get the count for a specific class, say class_id = 0
specific_class_count = class_counts[0]
print(f"Count for class 0: {specific_class_count}")
```

To export these counts, simply write them to a file:

```python
with open('class_counts.txt', 'w') as f:
    for class_id, count in class_counts.items():
        f.write(f"Class {class_id}: {count}\n")
```

Hope this helps! Let me know if you have any more questions.

**Reply by Jones034** [reply]
Hi, thank you very much for your feedback.
I was able to implement the export of the data :)
The problem with the counter is that I can't manage to find the correct output variable.

**Reply by glenn-jocher** [reply]
@Jones034 hi there! 😊

Great to hear you've successfully implemented the data export! For the counter issue, if you're using YOLOv8's Python API, the detection results including class IDs are stored in the `results` object. You can access the class IDs like this:

```python
class_ids = results.xyxy[0][:, -1].cpu().numpy()  # Assuming you're working with the first image in a batch
```

Then, you can count occurrences of each class ID using:

```python
from collections import Counter
class_counts = Counter(class_ids)
```

This gives you a dictionary where keys are class IDs and values are counts. To get the name of each class, you can use `results.names`.

Hope this helps! Let us know if you have any more questions.

**Comment by dgodus** [comment]
I have a question. I want to tell in the video or superimpose the objects that I have for example.
people: 6
dogs: 7
I want to implement in this code.
from ultralytics import YOLO
from ultralytics.solutions import object_counter
import cv2

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Define line points
line_points = [(20, 400), (1080, 400)]

# Video writer
video_writer = cv2.VideoWriter("object_counting_output.avi",
                       cv2.VideoWriter_fourcc(*'mp4v'),
                       fps,
                       (w, h))

# Init Object Counter
counter = object_counter.ObjectCounter()
counter.set_args(view_img=True,
                 reg_pts=line_points,
                 classes_names=model.names,
                 draw_tracks=True)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break
    tracks = model.track(im0, persist=True, show=False)

    im0 = counter.start_counting(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()

**Reply by pderrenger** [reply]
Hey there! 👋 Based on your request to count objects of different classes and then superimpose those counts onto the video, your current setup looks pretty good! To count and display separate object classes, you'd make a few adjustments within the `counter.start_counting` functionality or augment it to aggregate counts per class.

Since `tracks` likely contains class information for each tracked object, you'd iterate over them to increment counters for each class. Assuming `tracks` has a `cls` attribute or similar, here's a simplified example of how you might adjust your loop:

```python
# Before your loop, initialize counters for each class of interest
class_counters = {"person": 0, "dog": 0}

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break
    tracks = model.track(im0, persist=True, show=False)
    
    # Reset counters for each frame if counting per frame, or comment out if counting over the whole video
    # class_counters = dict.fromkeys(class_counters, 0)
    
    # Update counters based on detected classes in tracks
    for track in tracks:
        cls_name = model.names[track.cls]  # Assuming track has 'cls' attribute which is class ID
        if cls_name in class_counters:
            class_counters[cls_name] += 1
    
    # Optionally, sum up counts and draw on im0
    display_text = " ".join([f"{cls}: {count}" for cls, count in class_counters.items()])
    cv2.putText(im0, display_text, (20, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (255,255,255), 2, cv2.LINE_AA)
    
    video_writer.write(im0)

# Don't forget to reset counters here if you're counting across the entire video and need the final tallies
print(class_counters)  # For terminal output or further processing
```

Regarding exporting the counted values, after your while loop ends (right after `cap.release()`), you have several options. You can simply print them out as shown above, write them to a file (e.g., using `with open('counts.txt', 'w') as f: f.write(str(class_counters))`), or even integrate with a database or an analytics service depending on your requirements.

Hope this helps! Let me know if you have any more questions. 😊

**Reply by dgodus** [reply]
I didn't understand this part. Could you give me a more complete example.
         cls_name = model.names[track.cls] # Assuming track has 'cls' attribute which is class ID

**Reply by pderrenger** [reply]
Hey there! Sure, I'd be happy to clarify that for you! 😊

When you're working with the `tracks` returned from `model.track()`, each `track` object represents a detected and tracked object in the frame. The `track.cls` attribute is an integer representing the class ID of the detected object (e.g., 0 for a person, 1 for a bicycle, etc., depending on the model's training data).

`model.names` is a list where each index corresponds to a class ID and the value at each index is the human-readable string name of the class. So, `model.names[track.cls]` gives you the string name of the class detected by the tracker.

Here is an example to illustrate this:

```python
tracks = model.track(im0, persist=True, show=False)
for track in tracks:
    cls_id = track.cls  # This is the class ID (integer)
    cls_name = model.names[cls_id]  # This converts class ID to its string representation
    print(f'Track class ID: {cls_id}, Class name: {cls_name}')
```

This code snippet loops through each track in the `tracks` list, extracts the class ID, and then uses that ID to fetch the class's name from `model.names`, printing both the ID and the class name.

Hope this clears things up! If you have any more questions, feel free to ask.

**Reply by dgodus** [reply]
Could you give me a more complete example. please thank you for your response

**Reply by pderrenger** [reply]
@dgodus certainly! Let's dive into a more straightforward example for displaying counts of different classes on a video frame with YOLOv8:

```python
from collections import Counter
import cv2
from ultralytics import YOLO

# Load model
model = YOLO('yolov8n.pt')

# Video capture setup
cap = cv2.VideoCapture('path/to/video.mp4')
assert cap.isOpened(), 'Error opening video file'

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Perform tracking
    results = model.track(frame, persist=True)

    # Count occurrences of each class ID in detections
    class_ids = [track.cls for track in results]  # Assuming 'cls' gives the class ID
    counts = Counter(class_ids)

    # Prepare display text from counts
    display_text = ' '.join([f'{model.names[cls_id]}: {count}' for cls_id, count in counts.items()])

    # Display counts on the frame
    cv2.putText(frame, display_text, (50, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 2)

    # Show the frame
    cv2.imshow('Frame', frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()
```

This example simply tracks the objects in each video frame using `model.track()` and counts the occurrences of each object class. It then prepares a text string summarizing these counts and displays this text on each video frame.

Hope this example clarifies things for you! 😊 If you have any more questions, just let us know.

**Reply by dgodus** [reply]
when executing the cosa I have this error.
AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_115745/29624324.py in <module>
     19 
     20     # Count occurrences of each class ID in detections
---> 21     class_ids = [track.cls for track in results]  # Assuming 'cls' gives the class ID
     22     counts = Counter(class_ids)
     23 

/tmp/ipykernel_115745/29624324.py in <listcomp>(.0)
     19 
     20     # Count occurrences of each class ID in detections
---> 21     class_ids = [track.cls for track in results]  # Assuming 'cls' gives the class ID
     22     counts = Counter(class_ids)
     23 

~/.local/lib/python3.10/site-packages/ultralytics/utils/__init__.py in __getattr__(self, attr)
    154         """Custom attribute access error message with helpful information."""
    155         name = self.__class__.__name__
--> 156         raise AttributeError(f"'{name}' object has no attribute '{attr}'. See valid attributes below.\n{self.__doc__}")
    157 
    158 

AttributeError: 'Results' object has no attribute 'cls'. See valid attributes below.

**Reply by pderrenger** [reply]
@dgodus hey there! 😊 It looks like there's a small mix-up in accessing the tracked objects' class IDs. The attribute you're trying to access directly from the `Results` object isn't available in that way. Instead, you should iterate over `results.pred` to access each detection's properties, including the class ID (`cls`). Here's how you can adjust your code:

```python
class_ids = [det['cls'] for result in results for det in result.pred]
```

This approach will safely gather class IDs from your detections. If there's anything else I can help you with, feel free to reach out. Happy coding!

**Comment by dgodus** [comment]


**Reply by glenn-jocher** [reply]
Certainly! Here's a simplified way to count objects for two classes and export these values. Let's assume your two classes are 'Person' (class_id 0) and 'Car' (class_id 2).

First, after running inference with your model, loop through the detection results. Create two counters, one for each class, and increment them based on the class IDs detected:

```python
from ultralytics import YOLO

# Initialize counters
person_counter = 0
car_counter = 0

# Load model and run inference
model = YOLO('yolov8n.pt')  # or your custom model
results = model('path/to/image/or/video')

# Iterate through detections
for det in results.detections:
    if det.class_id == 0:  # Assuming '0' is the class ID for Person
        person_counter += 1
    elif det.class_id == 2:  # Assuming '2' is the class ID for Car
        car_counter += 1

print(f"Persons detected: {person_counter}")
print(f"Cars detected: {car_counter}")

# Exporting counted values
with open('counted_values.txt', 'w') as f:
    f.write(f"Persons detected: {person_counter}\n")
    f.write(f"Cars detected: {car_counter}\n")
```

This code snippet will print the detected counts to the console and also save them to a file named `counted_values.txt`. Please adjust the `class_id` values and the model path according to your actual use case. 🚗🚶‍♂️
